### PR TITLE
Terraform Provider Documentation Alignment

### PR DESCRIPTION
**Pull request by lornest**
Hey @jdamata! 

This is just a small change to align the documentation with the expected format for a Terraform Provider (see examples in the [Hashicorp generator](https://github.com/hashicorp/terraform-provider-scaffolding-framework/blob/main/docs/resources/example.md)).

This has no functional impact, but is not purely selfless - a number of tools that have been built around the TF ecosystem rely on these metadata - for example, I'm trying to run Upjet to generate a Kubernetes controller based off of your provider, which seems to scrape this metadata from the docs.

Thanks!

### CODE REVIEW

**Comment by tbutler-qontigo** [comment]
Nice idea...but how do we know that this is valid for said format...or indeed what stops the next PR just messing this up?
You are potentially creating a maintenance burden that nobody will respect because nobody really cares :)
Is there some sort of linting or other automated tool you could add that would run for PRs to validate that the docs are still compliant with the required format?

Otherwise it means we are taking your word for it that they are valid today and you are hoping that going forward, others will ensure any new/changed docs also comply

**Comment by lornest** [comment]
Sure, I can look at generating directly from this: https://github.com/hashicorp/terraform-plugin-docs and adding the relevant descriptions etc that are already in the hand cranked docs.

Could then look at using the `tfplugindocs validate` as part of PR actions?

**Comment by tbutler-qontigo** [comment]
That would make sense to me though @jdamata would have final say since it is his repo!
You might be able to take inspiration from https://github.com/terraform-docs/gh-actions which does for terraform modules what you want to do for terraform plugins.

**Comment by lornest** [comment]
Okay @tbutler-qontigo, @jdamata, I've now implemented tfplugindocs approach to generate documentation from the provider source code (resources and data sources), templates and examples.

I've taken the descriptions from the previous documentation to preserve as much as possible.

Where things were not possible to be generated from the source, I've used templates.

I've also implemented the `tfplugindocs-check` workflow. This checks that the documentation is up-to-date. It will fail if `make docs` is not run before committing, and something is added that makes the docs out of date. I've updated the `README.md` to add the need to run `make docs` before committing (NOTE: it's possible to add an 'on pull request' workflow for this, if you'd like me to take a look at this?)

Let me know your thoughts.

File: .DS_Store
Comment (jdamata) [inline .DS_Store:1]: can you add .DS_Store to the gitignore instead of committing this. Thanks!

File: .github/workflows/tfplugindocs-check.yml
Comment (jdamata) [inline .github/workflows/tfplugindocs-check.yml:5]: we dont have branches named release/**.

I think having this workflow check if a PR has generated docs correctly is sufficient

File: .github/workflows/tfplugindocs-check.yml
Comment (jdamata) [inline .github/workflows/tfplugindocs-check.yml:24]: Is there a reason we want to pin these actions to the git sha instead of just the semantic version? The comments is helpful but i would prefer just the version straight up if that works.

File: Makefile
Comment (jdamata) [inline Makefile:15]: are these rm's required? I'm guessing we have them here because `tfplugindocs` errors out when a file already exists?

File: .github/workflows/tfplugindocs-check.yml
Comment (jdamata) [inline .github/workflows/tfplugindocs-check.yml:1]: Can we shorten this name to just tfplugindocs check

**Comment by jdamata** [comment]
Thanks for the contribution @lornest. I added some small nitpick comments. 

All the documentation being shifted around is a lot to read through and verify its all good. Ill try to read through all the doc diffs and catch anything missed

File: .github/workflows/tfplugindocs-check.yml
Comment (freeranger) [reply .github/workflows/tfplugindocs-check.yml:24]: @jdamata I’d actually prefer to see this in all cases - I have seen two incidents recently where pinning to a major version caused issues.
- In one case, the action author did not respect semantic versioning and introduced a breaking change, which broke us
- In another case the action author accidentally deleted the version tag so the action “disappeared”

Using the commit SHA would have meant we were unaffected by these problems.

Using the SHA is also a recommended security hardening practice by github themselves: https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions#using-third-party-actions

Obviously your repo, your choice - but for me the above would be three good reasons to use the SHA

**Comment by freeranger** [comment]
> I've updated the `README.md` to add the need to run `make docs` before committing (NOTE: it's possible to add an 'on pull request' workflow for this, if you'd like me to take a look at this? - it requires the contributor setting a GitHub token secret to the workflow can generate the docs and commit them to their pull request if there are diffs)
> 
> Let me know your thoughts.

Great stuff @lornest but as you say, you have now introduced an additional step for the developer to have to run an manual step to generated automated documentation - it seems like it would be better to do as you suggest above - ie to have the PR do this step automatically.
That way the contributor has no extra steps to perform to get the documentation up to date.
I’m not sure what you mean about the contributor setting a token - isnt the built in token enough?

what do you think @jdamata

File: templates/debugging.md
Comment (freeranger) [inline templates/debugging.md:1]: Why have you introduced this template? `debugging.md` is not templated docs and not “part” of the provider as such - it is simply some information to allow anyone developing the provider to debug it.
If all docs need to become templates for some reason then you will need  to add something to the docs explaining this - otherwise someone may contribute another `.md` file, or indeed may make changes to the version in the docs folder because there does not seem to be any reason for the file to be templated.

If you want (or need) the files to be templates then I think you should add a [markdown comment](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#hiding-content-with-comments) at the top of every generated file so that any reader of the docs who may be considering editing them will know to edit the template and not the doc they are reading

Comment (freeranger) [inline templates/debugging.md:1]: Also, agai if this needs to be a template, why does it not need a header with page title etc as you have added elsewhere?

File: .github/workflows/tfplugindocs-check.yml
Comment (jdamata) [reply .github/workflows/tfplugindocs-check.yml:24]: I see. Good points, i can get behind using the git sha

**Comment by jdamata** [comment]
> > I've updated the `README.md` to add the need to run `make docs` before committing (NOTE: it's possible to add an 'on pull request' workflow for this, if you'd like me to take a look at this? - it requires the contributor setting a GitHub token secret to the workflow can generate the docs and commit them to their pull request if there are diffs)
> > Let me know your thoughts.
> 
> Great stuff @lornest but as you say, you have now introduced an additional step for the developer to have to run an manual step to generated automated documentation - it seems like it would be better to do as you suggest above - ie to have the PR do this step automatically. That way the contributor has no extra steps to perform to get the documentation up to date. I’m not sure what you mean about the contributor setting a token - isnt the built in token enough?
> 
> what do you think @jdamata

yah i agree and that would be sweet. @lornest is this something you could take a stab at?

File: .DS_Store
Comment (lornest) [reply .DS_Store:1]: Good catch, thanks

File: templates/debugging.md
Comment (lornest) [reply templates/debugging.md:1]: I added this template to maintain the position of `debuggin.md` in the current document structure. 

The generated files all have the metadata at the top containing `# generated by https://github.com/hashicorp/terraform-plugin-docs`.

perhaps I misunderstood the premise of this comment, please let me know if there's something else?

Comment (lornest) [reply templates/debugging.md:1]: `debugging.md` being in templates as a .md file and not .md.tmpl file just ensures it is copied as is. But it's a good suggestion to add the metadata/generated comment at the top for clarity, I'll do that now

**Comment by lornest** [comment]
@jdamata will address all of the above comments from you and the other contributors/maintainers and add the on PR workflow today. 

Thanks for the feedback everyone 👍

File: templates/debugging.md
Comment (freeranger) [reply templates/debugging.md:1]: In the first instance I was wondering why this needs to be a template at all and could not just have stayed where it was and not be a template at all.
The comment that says it is auto generated is all well and good…but generated from where?
Think about a new contributor to the repo -they spot a spelling mistake in `debugging.md` - the comment tells them this is autogenerated..but where do they go to find the original and update it?
Or someone wants to add a new documentation file to describe something else…where do they put that file?  If it needs to be a template, is it enough that it exists in the template folder or does something else need to happen?
What if they just put it in the folder they want it to appear in?

It is not particularly intuitive to have to go looking for templates for files that are essentially stand-alone documentation - if it is strictly necessary for them to be templates then you must consider from the point of view of someone who neither knows nor cares about the templating system - what is the least friction you can cause them to achieve what they want?   Guiding them where necessary and automating (as per my suggestion on generating docs at PR time rather than forcing the contributor to do a manual run) where possible helps to reduce friction

Comment (lornest) [reply templates/debugging.md:1]: Got you now @freeranger. I've changed this approach up to have `debugging.md` just remain in the `docs` directory, rather than being cloned as part of the generation. As such, I've also actioned @jdamata's point about the `rm` commands in the `make docs` target. This was why I had debugging in the templates - as make docs essentially removed the docs dir and generates each time to ensure they are up to date (slightly heavy handed, perhaps). I've tested now with the `debugging.md` in `/docs` and the target without `rm` commands and all seems well.

Comment (lornest) [reply templates/debugging.md:1]: Would you prefer I did the same with `index.md` @freeranger/ @jdamata  ? It also doesn't have any auto-generated content, as I just took the existing index file as it was.

Comment (lornest) [reply templates/debugging.md:1]: So `index.md` does need templated to maintain consistency with the existing docs - the generated docs lose some elements.

The difference with the header is that `debugging.md` isn't actually part of the 'site'. If you look on the provider docs it doesn't appear (https://registry.terraform.io/providers/jdamata/sonarqube/latest/docs). It seems it was moved into the `/docs` directory last year, but by default Terraform provider docs don't expect this. I could put it into `/docs/guides/debugging.md` with the metadata header, then it would show on the site. But is this what you want? Or should this debugging info live outside of `/docs` at the root of the repo like a `contributing.md` would?

Comment (freeranger) [reply templates/debugging.md:1]: Ok, fair enough about `index.md` 
I don’t think `debugging.md` needs to be part of the provider docs since it only applies to debugging the provider and not to using it.
Whether it should stay where it is and stay like it is or move to another folder I will leave to @jdamata.
I am happy to leave it where it is but don’t feel strongly enough about the location.
Thanks

Comment (lornest) [reply templates/debugging.md:1]: Thakns @freeranger for the feedback. Will hang off to see what @jdamata's thoughts are and work on the other tidy ups in the meantime.

Comment (jdamata) [reply templates/debugging.md:1]: Im cool with moving it to the root of the repo.

**Comment by lornest** [comment]
Hey @jdamata and co!

I've added 2 workflows and tested them on my local fork. 

First, when someone creates a PR with changes to the provider code (`/sonarqube/...`) or any files that are used in the generation of docs (templates, examples, makefile) then the action generates updated docs and pushes them to the PR branch.

There's also another workflow that checks the documents are up-to-date on push to `master`. This is pretty much belt-and-braces if all code gets into master via PRs, but it might be useful if anyone from the contributors team ever pushes to master directly.

Let me know if there's anything else!

File: .github/workflows/tfplugindocs-generate.yml
Comment (jdamata) [inline .github/workflows/tfplugindocs-generate.yml:18]: lets be consistent and always use the git SHA in these github actions ref. Similar to how you have done in the tfplugin-check.yml

**Comment by jdamata** [comment]
> Hey @jdamata and co!
> 
> I've added 2 workflows and tested them on my local fork.
> 
> First, when someone creates a PR with changes to the provider code (`/sonarqube/...`) or any files that are used in the generation of docs (templates, examples, makefile) then the action generates updated docs and pushes them to the PR branch.
> 
> There's also another workflow that checks the documents are up-to-date on push to `master`. This is pretty much belt-and-braces if all code gets into master via PRs, but it might be useful if anyone from the contributors team ever pushes to master directly.
> 
> Let me know if there's anything else!

small nit in a comment. otherwise LGTM

**Review by jdamata — APPROVED** [review]

File: .github/workflows/tfplugindocs-generate.yml
Comment (lornest) [reply .github/workflows/tfplugindocs-generate.yml]: Good catch @jdamata. I've updated these to use git SHA and retested on my fork to confirm everything still works as expected. 

Republished this PR - forgot I'd moved it back to draft whilst I was working on the changes.

**Comment by lornest** [comment]
Thanks all, I've completed the updates as per the feedback

**Pull request merged**
